packages/table/src/interactions/selectable.tsx

```diff
@@ -110,142 +124,92 @@ export class DragSelectable extends React.Component<IDragSelectableProps, {}> {
     }
 
     private handleActivate = (event: MouseEvent) => {
+        const { locateClick, selectedRegions, selectedRegionTransform } = this.props;
         if (this.shouldIgnoreMouseDown(event)) {
             return false;
         }
 
-        let region = this.props.locateClick(event);
+        let region = locateClick(event);
 
         if (!Regions.isValid(region)) {
             return false;
         }
 
-        const {
-            onSelection,
-            selectedRegions,
-            selectedRegionTransform,
-        } = this.props;
-
-        const focusCellCoordinates = Regions.getFocusCellCoordinatesFromRegion(region);
         if (selectedRegionTransform != null) {
             region = selectedRegionTransform(region, event);
         }
 
         const foundIndex = Regions.findMatchingRegion(selectedRegions, region);
-        if (foundIndex !== -1) {
-            // If re-clicking on an existing region, we either carefully
-            // remove it if the meta key is used or otherwise we clear the
-            // selection entirely.
-            if (DragEvents.isAdditive(event)) {
-                const newSelectedRegions = selectedRegions.slice();
-                newSelectedRegions.splice(foundIndex, 1);
-                onSelection(newSelectedRegions);
-            } else {
-                onSelection([]);
-            }
-            const fullFocusCellCoordinates: IFocusedCellCoordinates = {
-                col: focusCellCoordinates.col,
-                focusSelectionIndex: null,
-                row: focusCellCoordinates.row,
-            };
-            this.props.onFocus(fullFocusCellCoordinates);
+        const matchesExistingSelection = foundIndex !== -1;
 
+        if (matchesExistingSelection) {
+            this.handleUpdateExistingSelection(foundIndex, event);
```

Comment:
this whole function is much easier to read now IMO. i basically just pulled the logic out into clearly, consistently named helpers